src/Evolution/DiscontinuousGalerkin/Actions/NormalCovectorAndMagnitude.hpp

```diff
@@ -0,0 +1,181 @@
+  using type = Scalar<DataVector>;
+};
+
+template <size_t Dim>
+struct NormalVector : db::SimpleTag {
+  using type = tnsr::I<DataVector, Dim, Frame::Inertial>;
+};
+
+/*!
+ * \brief Computes the normal covector, magnitude of the unnormalized normal
+ * covector, and in curved spacetimes the normal vector.
+ */
+template <typename System, size_t Dim, typename FieldsOnFaceTags>
+void unit_normal_vector_and_covector(
+    const gsl::not_null<Scalar<DataVector>*> face_normal_magnitude,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        unit_normal_covector,
+    const gsl::not_null<Variables<FieldsOnFaceTags>*> fields_on_face,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        unnormalized_normal_covector) noexcept {
+  if constexpr (has_inverse_spatial_metric_tag_v<System>) {
+    using inverse_spatial_metric_tag =
+        typename System::inverse_spatial_metric_tag;
+    auto& normal_vector = get<NormalVector<Dim>>(*fields_on_face);
+    const auto& inverse_spatial_metric =
+        get<inverse_spatial_metric_tag>(*fields_on_face);
+    // Compute unnormalized normal vector
+    for (size_t i = 0; i < Dim; ++i) {
+      normal_vector.get(i) = inverse_spatial_metric.get(i, 0) *
+                             get<0>(unnormalized_normal_covector);
+      for (size_t j = 1; j < Dim; ++j) {
+        normal_vector.get(i) += inverse_spatial_metric.get(i, j) *
+                                unnormalized_normal_covector.get(j);
+      }
+    }
+    // Perform normalization
+    dot_product(face_normal_magnitude, normal_vector,
+                unnormalized_normal_covector);
+    get(*face_normal_magnitude) = sqrt(get(*face_normal_magnitude));
+    auto& one_over_normal_vector_magnitude =
+        get<OneOverNormalVectorMagnitude>(*fields_on_face);
+    get(one_over_normal_vector_magnitude) = 1.0 / get(*face_normal_magnitude);
+    for (size_t i = 0; i < Dim; ++i) {
+      unit_normal_covector->get(i) = unnormalized_normal_covector.get(i) *
+                                     get(one_over_normal_vector_magnitude);
+      normal_vector.get(i) *= get(one_over_normal_vector_magnitude);
+    }
+  } else {
+    magnitude(face_normal_magnitude, unnormalized_normal_covector);
+    auto& one_over_normal_vector_magnitude =
+        get<OneOverNormalVectorMagnitude>(*fields_on_face);
+    get(one_over_normal_vector_magnitude) = 1.0 / get(*face_normal_magnitude);
+    for (size_t i = 0; i < Dim; ++i) {
+      unit_normal_covector->get(i) = unnormalized_normal_covector.get(i) *
+                                     get(one_over_normal_vector_magnitude);
+    }
+  }
+}
+
+/*!
+ * \brief Computes the normal vector, covector, and their magnitude on the face
+ * in the given direction.
+ */
+template <typename System, size_t Dim, typename DbTagsList,
+          typename FieldsOnFaceTags>
+void unit_normal_vector_and_covector_and_magnitude(
+    const gsl::not_null<db::DataBox<DbTagsList>*> box,
```

Comment:
Isn't this precisely the kind of interface (some unknown box as input, unknown changes to the box) we learned to avoid from SpEC? As @moxcodes said, it seems like passing the two things you use from the box would suffice. I've we've missed something and this interface can't be avoided, then I suggest adding careful documentation about what tags must be in the box and what tags will be changed, to reduced the guaranteed future confusion.